src/Servers/Kestrel/test/InMemory.FunctionalTests/MaxRequestBodySizeTests.cs

```diff
@@ -60,6 +60,89 @@ public async Task RejectsRequestWithContentLengthHeaderExceedingGlobalLimit()
         Assert.Equal(CoreStrings.FormatBadRequest_RequestBodyTooLarge(globalMaxRequestBodySize), requestRejectedEx.Message);
     }
 
+    [Fact]
+    public async Task RejectsRequestWithBodySizeExceedingPerRequestLimitAndExceptionWasCaughtByApplication()
+    {
+        var maxRequestBodySize = 10;
+        await using (var server = new TestServer(async context =>
+        {
+#pragma warning disable CS0618 // Type or member is obsolete
+            BadHttpRequestException requestRejectedEx = await Assert.ThrowsAsync<BadHttpRequestException>(async () =>
+#pragma warning restore CS0618 // Type or member is obsolete
+            {
+                using (var stream = new StreamReader(context.Request.Body))
+                {
+                    string body = await stream.ReadToEndAsync();
+                }
+            });
+            context.Response.StatusCode = 413;
+            await context.Response.WriteAsync("");
+            throw requestRejectedEx;
+        },
+        new TestServiceContext(LoggerFactory) { ServerOptions = new KestrelServerOptions { Limits = { MaxRequestBodySize = maxRequestBodySize } } }))
+        {
+            using var connection = server.CreateConnection();
+            await connection.Send(
+                "POST / HTTP/1.1",
+                "Host:",
+                "Content-Length: " + (maxRequestBodySize + 1),
+                "",
+                new string('a', maxRequestBodySize + 1));
+            await connection.ReceiveEnd(
+                "HTTP/1.1 413 Payload Too Large",
+                "Connection: close",
+                $"Date: {server.Context.DateHeaderValue}",
+                "Server: Kestrel",
+                "Transfer-Encoding: chunked",
+                "",
+                "");
+        }
+    }
+
+    [Fact]
+    public async Task RejectsRequestWithChunckedBodySizeExceedingPerRequestLimitAndExceptionWasCaughtByApplication()
+    {
+        var maxRequestBodySize = 10;
+        var chunkedPayload = $"5;random chunk extension\r\nHello\r\n6\r\n World\r\n0\r\n";
+        await using (var server = new TestServer(async context =>
+        {
+#pragma warning disable CS0618 // Type or member is obsolete
+            BadHttpRequestException requestRejectedEx = await Assert.ThrowsAsync<BadHttpRequestException>(async () =>
+#pragma warning restore CS0618 // Type or member is obsolete
+            {
+                using (var stream = new StreamReader(context.Request.Body))
+                {
+                    string body = await stream.ReadToEndAsync();
+                }
+            });
+            context.Response.StatusCode = 413;
```

Comment:
There is a better way to do this without direct mentioning 413? Here the intention is just forward the 413 to the response.

Reply:
You could use the [StatusCodes](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.statuscodes?view=aspnetcore-6.0) class if that's what you're asking. This has me thinking it would be better if these tests set a different status code and/or wrote a body to make it extra clear this is a custom response instead of the default 413 response. Although the `Transfer-Encoding: chunked` in the response does already subtly indicate that.

Reply:
no, is more about just using the original statuscode of the exception, like the Developer Exception Page does. Something like that: `context.Response.StatusCode = requestRejectedEx.StatusCode;`

Another thing that I'm worrying about is... if the some middleware of the application catches the `BadHttpRequestException` and wants to change the *final status code to 500* for example, it's still fine that the Kestrel send the `Connection: close` header?

Reply:
> no, is more about just using the original statuscode of the exception, like the Developer Exception Page does. Something like that: `context.Response.StatusCode = requestRejectedEx.StatusCode;`

Yep. Looks like you've already done it. You can do so without the `#pragma warning disable CS0618` if you use `Microsoft.AspNetCore.Http.BadHttpRequestException` instead of `Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException`, but it's not a big deal. The rest of the test class still uses the deprecated version, and updating everything right now isn't necessary.

> Another thing that I'm worrying about is... if the some middleware of the application catches the `BadHttpRequestException` and wants to change the *final status code to 500* for example, it's still fine that the Kestrel send the `Connection: close` header?

This is fine. Kestrel is going to close the connection either way if the previous request body payload was too large no matter the status code sent by the application. This merely indicates this intention ahead of time to the client.